### ISSUE TITLE
Resume is not bit-exact: AdamW optimizer state not restored to full precision

### ISSUE THREAD

**Issue by amazloumi** [KempnerInstitute/KempnerForge#114]
## What happened
Resuming from a DCP checkpoint is **state-correct but not bit-exact**. A resumed run does not reproduce the loss trajectory of an otherwise-identical uninterrupted run — even though uninterrupted same-seed runs ARE bit-deterministic (verified: max|Δloss|=0 across 1-GPU / FSDP / EP).

Isolation shows model weights, data-sample position, scheduler, and RNG are restored exactly, but the **AdamW optimizer moments are not**: the first optimizer step after resume diverges, then propagates.

Evidence (tiny dense model, real data, single GPU, bf16, fused AdamW; stop at step 8, resume to 16):
- Pre-resume steps 1–8: interrupted run **bit-identical** to reference → checkpoint-save is side-effect-free; the checkpoint captures the true state.
- Step 9 (first post-resume) `grad_norm` = **0.855 == 0.855** (ref == resumed) → model weights + data restored exactly.
- Step 10 `grad_norm`: ref **0.930** vs resumed **0.973** → the step-9 optimizer update differed → optimizer moments not bit-exact.
- Loss matches steps 9–11, then drifts ≤0.0625 (≈1 bf16 ULP) from step 12.

  ## Reproduction
  Reference (uninterrupted 16 steps) vs interrupt-at-8 + resume-to-16, comparing post-resume losses. Full script in our study worktree: `spikes/resume_test/verify_resume.sh`. Core:
  reference:           --train.max_steps=16 --checkpoint.dir=/tmp/ref  --checkpoint.interval=100
  
  interrupt at 8:      --train.max_steps=8  --checkpoint.dir=/tmp/test --checkpoint.interval=4

  resume to 16:        --train.max_steps=16 --checkpoint.dir=/tmp/test --checkpoint.interval=4

  (debug.toml, dim=128 n_layers=2 vocab=50304, bf16, compile off, seed 7, log_interval=1, real .npy data)

## Observed divergence 
  | step | ref grad_norm | resumed grad_norm | ref loss | resumed loss |
  |---|---|---|---|---|
  | 9  | 0.855 | 0.855 | 10.7500 | 10.7500 |
  | 10 | 0.930 | **0.973** | 10.6875 | 10.6875 |
  | 12 | 1.023 | 0.980 | 10.6250 | **10.5625** |
  | 16 | — | — | 10.5000 | **10.4375** |
  
## Environment
- GPU: NVIDIA H200 (single GPU — reproduces without distributed)
- PyTorch: 2.11.0+cu128
- KempnerForge: commit `2b08f8e`
- bf16 mixed precision, `optimizer.fused=true` (default), AdamW
  
## Likely cause / to investigate in the fix
Optimizer-state round-trip loses precision (AdamW `exp_avg`/`exp_avg_sq`, and/or fp32 master weights under bf16+fused). Check: does `optimizer.fused=false` fix it?
Are fp32 master weights checkpointed? DCP save/load dtype of the optimizer state_dict.
  
## Missing test (add in the fix PR)
No test asserts bit-exact resume — `test_{checkpoint,pp_checkpoint,moe_checkpoint}_resume` only check loss continuity / loss>0, so this passes today. Add a **resume-determinism test** (reference vs interrupt+resume, assert post-resume losses bit-identical) on **single GPU** and **distributed (2-GPU FSDP)**.
